### PR TITLE
Keep constrained dofs in composition block

### PR DESCRIPTION
**Pull request by gassmoeller**
@anne-glerum: First baby step towards being able to disable boundary conditions for some compositional fields by keeping constrained dofs in the sparsity pattern. @tjhei and @bangerth is this what we talked about on Tuesday?

### CODE REVIEW

File: source/simulator/core.cc [hunk 1]
```diff
@@ -1046,14 +1046,65 @@ namespace aspect
                                               face_coupling,
                                               Utilities::MPI::
                                               this_mpi_process(mpi_communicator));
+
+        if (solver_scheme_solves_advection_equations(parameters)
+            &&
+            compositional_fields_need_matrix_block(introspection))
+          {
+            // If we solve for more than one compositional field make sure we keep constrained entries
+            // to allow different boundary conditions for different fields.
```
Comment (bangerth) [inline source/simulator/core.cc:1055]: It took me a good long while to understand why you don't just merge the call to `make_flux_sparsity_pattern()` into the one above. Can you add a comment here or at the actual call site below mentioning that the point is that in the call here, "keeping constrained entries" implies passing `true` as fourth argument?

File: source/simulator/core.cc [hunk 2]
```diff
@@ -1046,14 +1046,65 @@ namespace aspect
...
-                                       this_mpi_process(mpi_communicator));
+      {
+        DoFTools::make_sparsity_pattern (dof_handler,
+                                         coupling, sp,
+                                         current_constraints, false,
+                                         Utilities::MPI::
+                                         this_mpi_process(mpi_communicator));
+
+        // If we solve for more than one compositional field make sure we keep constrained entries
+        // to allow different boundary conditions for different fields.
```
Comment (bangerth) [inline source/simulator/core.cc:1085]: Same here

**Review by bangerth — COMMENTED** [review]
OK with two small additions to the commentary.

**Comment by gassmoeller** [comment]
I had to spend a few minutes to remind myself again. I extended the comment.

**Review by bangerth — APPROVED** [review]

**Pull request merged**